File: client-android/overlays/Compliance/GDPRExportButton.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Mega Promoting SRL
//
// GDPR Article 15 + Article 20 self-service trigger. Calls
// POST /api/v1/users/me/export on the AI layer and presents the returned
// signed download URL.

package eu.mdchat.app.compliance

import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.FileUpload
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp

@Composable
fun GDPRExportButton(
    isExporting: Boolean,
    onExport: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Button(
        onClick = onExport,
        enabled = !isExporting,
        modifier = modifier,
    ) {
        Row {
            Icon(imageVector = Icons.Filled.FileUpload, contentDescription = null)
            Text(
                text = "Export my data",
                modifier = Modifier.padding(start = 8.dp),
            )
            if (isExporting) {
                CircularProgressIndicator(
                    modifier = Modifier
                        .padding(start = 8.dp)
                        .size(16.dp),
                    color = MaterialTheme.colorScheme.onPrimary,
                    strokeWidth = 2.dp,
                )
            }
        }
    }
}

private fun Modifier.size(size: androidx.compose.ui.unit.Dp): Modifier =
    this.then(androidx.compose.foundation.layout.size(size))

@Preview
@Composable
private fun GDPRExportButtonPreview() {
    MaterialTheme {
        GDPRExportButton(isExporting = false, onExport = {})
    }
}

File: client-ios/overlays/Compliance/GDPRExportButton.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Mega Promoting SRL
//
// GDPR Article 15 (access) + Article 20 (portability) self-service trigger.
// Calls the AI-layer REST endpoint `/api/v1/users/me/export` and presents the
// returned JSON file in a share sheet. SLA 30 days; this endpoint returns
// a same-day signed download URL.

import SwiftUI

public struct GDPRExportButton: View {

    @State private var isExporting = false
    @State private var lastError: String?
    @Environment(\.openURL) private var openURL

    public init() {}

    public var body: some View {
        Button {
            Task { await exportData() }
        } label: {
            HStack {
                Image(systemName: "square.and.arrow.up")
                Text(NSLocalizedString("export_my_data", comment: "GDPR Art 15 trigger"))
                if isExporting { ProgressView().padding(.leading, 8) }
            }
        }
        .disabled(isExporting)
        .alert("Export failed",
               isPresented: .constant(lastError != nil),
               actions: {
                   Button("OK") { lastError = nil }
               },
               message: { Text(lastError ?? "") })
    }

    private func exportData() async {
        isExporting = true
        defer { isExporting = false }

        // Replace baseURL with injected configuration in real wiring.
        guard let url = URL(string: "https://msg.md-chat.eu/api/v1/users/me/export") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
                lastError = "Server returned \((response as? HTTPURLResponse)?.statusCode ?? -1)."
                return
            }
            // Real wiring will parse the signed URL from response body + present share sheet.
            if let signed = http.allHeaderFields["X-Download-URL"] as? String, let signedURL = URL(string: signed) {
                openURL(signedURL)
            }
        } catch {
            lastError = error.localizedDescription
        }
    }
}

#Preview {
    GDPRExportButton()
        .padding()
}
